libraries/AC_AttitudeControl/AC_WeatherVane.cpp

```diff
@@ -0,0 +1,247 @@
+/*
+ * Aircraft Weathervane options common to vtol plane and copters
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AC_WeatherVane.h"
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Terrain/AP_Terrain.h>
+
+#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
+    #define WVANE_PARAM_ENABLED 1
+#else
+    #define WVANE_PARAM_ENABLED 0
+#endif
+
+
+const AP_Param::GroupInfo AC_WeatherVane::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable
+    // @Description{Copter}: Enable weather vaning.  When active, and the appropriate _OPTIONS bit is set for auto, or guided, the aircraft will yaw into wind once the vehicle is in a condition that meets that set by the WVANE parameters and there has been no pilot input for 3 seconds.
+    // @Description{Plane}: Enable weather vaning.  When active, the aircraft will automatically yaw into wind when in a VTOL position controlled mode. Pilot yaw commands overide the weathervaning action.
+    // @Values: 0:Disabled,1:Nose into wind,2:Nose or tail into wind,3:Side into wind
+    // @User: Standard
+    AP_GROUPINFO_FLAGS("ENABLE", 1, AC_WeatherVane, _direction, WVANE_PARAM_ENABLED, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: GAIN
+    // @DisplayName: Weathervaning gain
+    // @Description: This converts the target roll/pitch angle of the aircraft into the correcting (into wind) yaw rate. e.g. Gain = 2, roll = 30 deg, yaw rate = 60 deg/s.
+    // @Range: 0 4
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("GAIN", 2, AC_WeatherVane, _gain, 0.5),
+
+    // @Param: ANG_MIN
+    // @DisplayName: Weathervaning min angle
+    // @Description: The minimum target roll/pitch angle before active weathervaning will start.  This provides a dead zone that is particularly useful for poorly trimmed quadplanes.
+    // @Units: deg
+    // @Range: 0 10
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("ANG_MIN", 3, AC_WeatherVane, _min_dz_ang_deg, 1),
+
+    // @Param: ALT_MIN
+    // @DisplayName: Weathervaning min altitude
+    // @Description: Above this altitude weathervaning is permitted.  If terrain is enabled, AGL is used.  Set zero to ignore altitude.
+    // @Units: m
+    // @Range: 0 50
```

Comment:
The ground also has a negative altitude.
In Japan, there is ground with negative altitude.

Reply:
I have clarified the wording in the metadata.  This is actually a height not an altitude so that was my misuse of the word.  If a user wishes to fly close to the ground in locations other than home and have the weathervane activation dependant on that height, then they should enable terrain.